### ISSUE TITLE
PCSS not working for SpotLight3D: Vulkan (Forward+)

### ISSUE THREAD

**Issue by kevancress** [godotengine/godot#80137]
### Godot version

Godot v4.1.stable

### System information

Godot v4.1.stable - Windows 10.0.22621 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3060 Ti (NVIDIA; 31.0.15.2824) - 11th Gen Intel(R) Core(TM) i5-11600K @ 3.90GHz (12 Threads)

### Issue description

Percentage-closer soft shadows (PCSS) don't seem to be working correctly with SpotLight3d's.

Image below shows an Omni Light and Spot Light with the same light_size value. The Omni Light shows correct PCSS with the shadow getting softer the furth the object (caster) is from the surface (reciever). The Spot Light shows uniform shadow hardness.

![image](https://github.com/godotengine/godot/assets/16860574/a66cde9b-3ea2-4f3a-bada-e5654a1317b0)

Increasing the light_size Value for a Spot Light uniformly blurs the shadow, but the blur doesn't seem impacted by the casters distance from the receiver.

![image](https://github.com/godotengine/godot/assets/16860574/5ba77d2c-05c9-4e52-82b3-dac9b5aab439)

### Steps to reproduce

1. Create a 3D Scene 
2. Add a SpotLight3D
3. Add a plane mesh instance, and a shadow caster mesh instance
4. Enable shadows on the SpotLight3D
5. Change the light_size to 0.1
6. Observe
7. Change SpotLight3D to OmniLight3D with the same light_size
8. Observe

### Minimal reproduction project

[PCSS_Shadow_Issue.zip](https://github.com/godotengine/godot/files/12233905/PCSS_Shadow_Issue.zip)

**Comment by Calinou** [comment]
I can confirm this on 4.2.dev 41a7f6b38 (Linux, GeForce RTX 4090 with NVIDIA 535.54.03).

**Comment by jsjtxietian** [comment]
I can help with this issue if no one started working on it.

**Comment by jsjtxietian** [comment]
The problem lies in the different calculation method for `light_data.soft_shadow_size` between omni light and spot light
For omni light, just as simple as `light_data.soft_shadow_size = size;`
But for spot light, 
```
float half_np = cm.get_z_near() * Math::tan(Math::deg_to_rad(spot_angle));
light_data.soft_shadow_size = (size * 0.5 / radius) / (half_np / cm.get_z_near()) * rect.size.width; 
```
and essentially equals to ` ( size * 0.5 * rect.size.width ) / ( radius * Math::tan(Math::deg_to_rad(spot_angle) ))`
In the example provided, let's say origin soft_shadow_size is 0.5, rect.size.width is 0.25, radius is 5, and the final result is 0.0125.

**Comment by aelspire** [comment]
Maybe some clue:

Spot with range 100:

![Image](https://github.com/user-attachments/assets/893e1bff-11d8-4c69-9c68-60fe853102e3)

Spot with range 20:

![Image](https://github.com/user-attachments/assets/dd83ee10-a459-42ad-bf10-f28f79288125)

Spot with range 15:

![Image](https://github.com/user-attachments/assets/dabf89a5-0849-4332-930c-df7d716567f5)

In calculations:

``` cpp
float half_np = cm.get_z_near() * Math::tan(Math::deg_to_rad(spot_angle));
light_data.soft_shadow_size = (size * 0.5 / radius) / (half_np / cm.get_z_near()) * rect.size.width;
```

radius is:

``` cpp
float radius = MAX(0.001, light->param[RS::LIGHT_PARAM_RANGE]);
```

I'm rendering newbie so take it with grain of salt, but I don't think `light_data.soft_shadow_size` should depend on a spot light's range.

**Comment by clayjohn** [comment]
@aelspire Good Catch! That entire calculation is a bit fishy. `half_np` is immediately divided by `cm.get_z_near()` so `cm.get_z_near()` is effectively canceled out.

**Comment by kevancress** [comment]
Did some tinkering today, Looks like removing the relationship between the spotlights range and the` light_data.soft_shadow_size`   makes the PCSS soft shadows work as expected:

Before: 

![Image](https://github.com/user-attachments/assets/b92aef81-bb0d-4ae0-9b65-085eac976bce)

After:

![Image](https://github.com/user-attachments/assets/f9acc588-0697-4aa9-9d59-9591578af703)

Changes made are in:
https://github.com/kevancress/godot/commit/f5ad8c031d3eadf7fc1f24084ed056c8f41afc10

Happy to submit a PR for this if it makes sense to do so?

I did also change the z_norm calculation to be `1.0 - dot(spot_dir, -light_rel_vec)` , noticed that the z_norm calculation for omni lights  also swapped from `<somthing>` to `1.0 - <somthing>`  after 4.3 and am assuming this may something to do with the Reverse Z change around that time?

**Comment by kevancress** [comment]
hmm... still something not quite right about the changes I've made above, odd artifacts at other angles, will keep tinkering...

![Image](https://github.com/user-attachments/assets/559a78ac-48c1-4c36-a5bc-c1083300bfd5)
![Image](https://github.com/user-attachments/assets/f1fd51b3-b2b6-4342-83fd-b6a355ec17be)

**Comment by PIXEL-GAME-DEV** [comment]
I am in the process of creating a PC monitor and added a spot light for realistic lighting and this exact bug is present. I can confirm that the lights range affects the blur size.

**Comment by PIXEL-GAME-DEV** [comment]
After some more tinkering it turns out the shadows blur is inverted. Real shadows get blurrier the closer the object is and sharper the further it is due to the lights angular size from the surface.
In Godot it is opposite and spotlight shadows get blurrier the further the object is.
This is easily demonstrated by creating an OmniLight3D and SpotLight3D, setting both of them to the same size, and moving them back and forth next to a shadow casting object, taking note of the shadows penumbra.
OmniLight3D is working correctly but SpotLight3D is inverted.
This looks like an incorrect usage of the depth map values in SpotLight3D.
